pyro/ops/special.py

```diff
@@ -0,0 +1,67 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import functools
+import math
+import operator
+
+
+def log_beta(x, y):
+    """
+    Log Beta function.
+
+    :param torch.Tensor x: A positive tensor.
+    :param torch.Tensor y: A positive tensor.
+    """
+    return x.lgamma() + y.lgamma() - (x + y).lgamma()
+
+
+def log_beta_stirling(x, y, tol=0.1):
+    """
+    Shifted Stirling's approximation to the log Beta function.
+
+    This is useful as a cheaper alternative to :func:`log_beta`.
+
+    This adapts Stirling's approximation of the log Gamma function::
+
+        lgamma(z) ≈ (z - 1/2) * log(z) - z + log(2 * pi) / 2
+
+    to approximate the log Beta function::
+
+        log_beta(x, y) ≈ ((x-1/2) * log(x) + (y-1/2) * log(y)
+                          - (x+y-1/2) * log(x+y) + log(2*pi)/2)
+
+    This additionally improves accuracy near zero by iteratively shifting
+    the log Gamma approximation using the recursion::
+
+        lgamma(x) = lgamma(x + 1) - log(x)
+
+    If this recursion is applied ``n`` times, then absolute error is bounded by
+    ``error < 0.082 / n < tol``, thus we choose ``n`` based on the user
+    provided ``tol``.
+
+    :param torch.Tensor x: A positive tensor.
+    :param torch.Tensor y: A positive tensor.
+    :param float tol: Bound on maximum absolute error. Defaults to 0.1. For
+        very small ``tol``, this function simply defers to :func:`log_beta`.
+    """
+    assert isinstance(tol, (float, int)) and tol >= 0
+    if tol < 0.02:
+        # Eventually it is cheaper to defer to torch.lgamma().
+        return log_beta(x, y)
+
+    # This bound holds for arbitrary x,y. We could do better with large x,y.
+    shift = int(math.ceil(0.082 / tol))
+
+    xy = x + y
+    factors = []
+    for _ in range(shift):
+        factors.append(xy / (x * y))
+        x = x + 1
+        y = y + 1
+        xy = xy + 1
+
+    log_factor = functools.reduce(operator.mul, factors).log()
+
+    return (log_factor + (x - 0.5) * x.log() + (y - 0.5) * y.log()
+            - (xy - 0.5) * xy.log() + (math.log(2 * math.pi) / 2 - shift))
```

Comment:
Where does this `shift` come from?

Reply:
Stirling's approximation of each of the `lgamma` terms is
```
lgamma(x) ~ (x-1/2) log(x) - x + log(2 pi)/2
```
Now when we use this to approximate a `log_beta` function the `x` terms usually cancel
```
log_beta(x,y) = lgamma(x) + lgamma(y) - lgamma(x+y)
              = (x-1/2) log(x) + (y-1/2) log(y) - (x+y-1/2) log(x+y)
              + -x + -y + (x+y)       # <--- these cancel if shift = 0
              + (1+1-1) log(2 pi)/2
```
Now the trick to make this more accurate is to use the `lgamma` recursion and increment each of `x`, `y`, and `x+y`, which is written in the code as
```
x = x + 1
y = y + 1
xy = xy + 1
```
When we do that, the first line works unchanged, the second line gets an extra `-1`, and the last line remains unchanged. And if we increment each of `x`, `y`, and `x+y` `shift`-many times and each time gets an extra `-1` then we can simply add the term `shift`.

Reply:
Oh yeah, it comes from the second line (I thought it cancelled out). Thanks for a clear explanation!